desktop_version/src/Music.cpp

```diff
@@ -92,81 +131,263 @@ class SoundTrack
+        } else {
+            faudio_buffer.PlayLength = std::min(stb_vorbis_get_samples_float_interleaved(t->vorbis, t->channels, (float*)t->decoded_buf_reserve, t->size/sizeof(float)), (t->loopbegin+t->looplength)-t->sample_pos);
+        }
+        faudio_buffer.AudioBytes = t->size;
+        faudio_buffer.pAudioData = t->decoded_buf_reserve;
+        faudio_buffer.pContext = t;
+        if (faudio_buffer.PlayLength == 0) {
+            if (t->shouldloop) {
+                stb_vorbis_seek(t->vorbis, t->loopbegin);
+                t->sample_pos = t->loopbegin;
+                if (t->looplength != 0) {
+                    faudio_buffer.PlayLength = std::min(stb_vorbis_get_samples_float_interleaved(t->vorbis, t->channels, (float*)t->decoded_buf_reserve, t->size/sizeof(float)), t->looplength);
+                } else {
+                    faudio_buffer.PlayLength = stb_vorbis_get_samples_float_interleaved(t->vorbis, t->channels, (float*)t->decoded_buf_reserve, t->size/sizeof(float));
+                }
+                if (faudio_buffer.PlayLength == 0) return;
+            } else {
+                return;
+            }
+        }
+        t->sample_pos += faudio_buffer.PlayLength;
+        FAudioSourceVoice_SubmitSourceBuffer(musicVoice, &faudio_buffer, NULL);
+    }
+    static void swapBuffers(FAudioVoiceCallback* callback, void* ctx) {
+        MusicTrack* t = (MusicTrack*)ctx;
+        std::swap(t->decoded_buf_playing, t->decoded_buf_reserve);
+    }
+
+    //Lifted from sdl_mixer
+    static void parseComments(MusicTrack* t, char** comments, int comment_list_length) {
+        int loopend = 0;
+        for (int i = 0; i < comment_list_length; i++) {
+            char *param = SDL_strdup(comments[i]);
+            char *argument = param;
+            char *value = SDL_strchr(param, '=');
+            if (value == NULL) {
+                value = param + SDL_strlen(param);
+            } else {
+                *(value++) = '\0';
+            }
 
-static std::vector<SoundTrack> soundTracks;
-static std::vector<MusicTrack> musicTracks;
+            /* Want to match LOOP-START, LOOP_START, etc. Remove - or _ from
+            * string if it is present at position 4. */
+            //_Mix_IsLoopTag inlined
+            char buf[5];
+            SDL_strlcpy(buf, argument, 5);
+            if (SDL_strcasecmp(buf, "LOOP") == 0 && ((argument[4] == '_') || (argument[4] == '-'))) {
+                SDL_memmove(argument + 4, argument + 5, SDL_strlen(argument) - 4);
+            }
+
+            if (SDL_strcasecmp(argument, "LOOPSTART") == 0)
+                t->loopbegin = _Mix_ParseTime(value, t->format.nSamplesPerSec);
+            else if (SDL_strcasecmp(argument, "LOOPLENGTH") == 0) {
+                t->looplength = SDL_strtoll(value, NULL, 10);
+            } else if (SDL_strcasecmp(argument, "LOOPEND") == 0) {
```

Comment:
Closing brace needs to be on a separate line. Opening brace needs to be on the next line.